### PR TITLE
Rework plugin config system #2

### PR DESCRIPTION
**Pull request by DerEchtePilz**
Small follow-up for #596 because I was too quick

### CODE REVIEW

File: commandapi-platforms/commandapi-bukkit/commandapi-bukkit-test/commandapi-bukkit-test-tests/src/test/java/dev/jorel/commandapi/test/ConfigGenerationTests.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+	// Test methods
+	public void validateConfigOptions(Set<String> options, BukkitConfigurationAdapter adapter) {
+		boolean containsAll;
+		for (String option : options) {
+			containsAll = adapter.contains(option);
+			if (!containsAll) {
+				throw new IllegalStateException("Config option '" + option + "' does not exist!");
+			}
+		}
```
Comment (willkroboth) [inline commandapi-platforms/commandapi-bukkit/commandapi-bukkit-test/commandapi-bukkit-test-tests/src/test/java/dev/jorel/commandapi/test/ConfigGenerationTests.java:167]: Since we're in the test environment, using the methods from `org.junit.jupiter.api.Assertions` here and in the other `validate...` methods would match the other tests. I think the `AssertionFailedError`s thrown by `org.junit.jupiter.api.Assertions` also make it clearer if the tests fail that it was an unexpected result in the tests, and not a general `IllegalStateException` thrown by the production code.

```suggestion
		for (String option : options) {
			assertTrue(adapter.contains(option), "Config option '" + option + "' does not exist!");
		}
```

**Review by willkroboth — CHANGES_REQUESTED** [review]
Just one bug on Velocity and some general comments about the tests.

Neither of these are big deals, but I noticed two ways that `ConfigGenerationTests.java` has different general formatting to other test files. First the test class and test methods here have `public` visibility, while the other classes have no visibility keyword. That's just a JUnit convention :shrug:; if they don't need to be public, they get package visibility. SonarCloud complains about this :P

image

Second, tests like `CommandExecutionTests` or `CommandNamespaceTests` that use utility methods put those utilities above the tests. Here that would be all the `validate...` methods. I think utility methods on top makes sense? Like I'm reading top to bottom, I see the methods defined, then I see how they are used. The first time I read through `ConfigGenerationTests` I was a bit confused about what methods the tests were referencing. Not a big deal though because once I got to the bottom I knew what I was looking at and understood on my second read.


To make the coverage reports work, I had to add these dependencies to `commandapi-codecov`:

```xml
<dependency>
	<groupId>dev.jorel</groupId>
	<artifactId>commandapi-bukkit-plugin-common</artifactId>
	<version>${project.version}</version>
</dependency>
<dependency>
	<groupId>dev.jorel</groupId>
	<artifactId>commandapi-plugin</artifactId>
	<version>${project.version}</version>
</dependency>
```

From the coverage, tests seem pretty good. There's just a few branches missed in `BukkitConfigurationAdapter#tryCreateSection`:

image

`ConfigGenerator#generate` never returned `null`:

image

`BukkitConfigurationAdapter#saveDefaultConfig` was never run at all, and the default `DefaultBukkitConfig` was never created:

image

I think `YamlConfiguration#saveToString` could be helpful for testing those last two things. Maybe could check that saving the default config to file results in the correct String? Instead of including the full expected String in the test file directly, the expected config String could be loaded from another file.

**Comment by DerEchtePilz** [comment]
> First the test class and test methods here have `public` visibility, while the other classes have no visibility keyword. That's just a JUnit convention 🤷; if they don't need to be public, they get package visibility.

Ah, okay, wasn't aware of that. Will change.

> To make the coverage reports work, I had to add these dependencies to `commandapi-codecov`

Yeah, I am aware of that and it exists as a local change already.

> `ConfigGenerator#generate` never returned `null`

A test for that exists locally already.

File: commandapi-platforms/commandapi-velocity/commandapi-velocity-plugin/src/main/java/dev/jorel/commandapi/config/VelocityConfigurationAdapter.java [hunk 2]
```diff
@@ -102,9 +87,9 @@ public void saveDefaultConfig(File directory, File configFile, Logger logger) {
 				logger.severe("Failed to create directory for the CommandAPI's config.yml file!");
 			}
 			try {
-				ConfigurationAdapter<ConfigurationNode> velocityConfigurationAdapter = new VelocityConfigurationAdapter(configLoader, configLoader.createNode(), defaultConfig);
-				configGenerator.populateDefaultConfig(velocityConfigurationAdapter);
-				configLoader.save(velocityConfigurationAdapter.config());
+				ConfigurationAdapter<ConfigurationNode> velocityConfigurationAdapter = new VelocityConfigurationAdapter(loader, loader.createNode(), defaultConfig);
+				configGenerator.generate(velocityConfigurationAdapter);
+				loader.save(velocityConfigurationAdapter.config());
```
[the remark this answers is not in the document]
Comment (DerEchtePilz) [reply commandapi-platforms/commandapi-velocity/commandapi-velocity-plugin/src/main/java/dev/jorel/commandapi/config/VelocityConfigurationAdapter.java:92]: > Although, I still think merging this logic between Bukkit and Velocity in `ConfigGenerator` would help avoid issues like this where the code that should do the same thing is correct in one version but has a bug in the other.

I know you want me to put the `saveDefaultConfig()` method into the `ConfigGenerator` class, however as mentioned already, I *really* don't think that that method is correct in there and I **really** don't want to put it in there. Although, I have another idea to avoid issues like this.

File: commandapi-platforms/commandapi-bukkit/commandapi-bukkit-test/commandapi-bukkit-test-tests/src/test/java/dev/jorel/commandapi/test/ConfigGenerationTests.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+	}
+
+	private void validateSections(List<String> sections, String expectedOption, YamlConfiguration config) {
+		ConfigurationSection root = config.getConfigurationSection(sections.get(0));
+		assertNotNull(root, "Section '" + sections.get(0) + "' does not exist!");
+
+		for (int i = 1; i < sections.size(); i++) {
+			root = root.getConfigurationSection(sections.get(i));
+			assertNotNull(root, "Section '" + sections.get(i) + "' does not exist!");
+		}
```
Comment (willkroboth) [inline commandapi-platforms/commandapi-bukkit/commandapi-bukkit-test/commandapi-bukkit-test-tests/src/test/java/dev/jorel/commandapi/test/ConfigGenerationTests.java:115]: The case for the first element could be included in the loop, but it works the same :shrug:.

```suggestion
		ConfigurationSection root = config;
		for (String section : sections) {
			root = root.getConfigurationSection(section);
			assertNotNull(root, "Section '" + section + "' does not exist!");
		}
```

**Review by willkroboth — COMMENTED** [review]
Still some parts of the code untested: the branches in `BukkitConfigurationAdapter#tryCreateSection`, the `saveDefaultConfig` method, and the rest is stuff `saveDefaultConfig` calls. Those methods look good but the edge cases are important for the functionality, so I think it would be nice to get them covered in tests.
Also, while `BukkitConfigruationAdapter#complete` is called, its effect of placing `\n` in the right places isn't verified, as the string version of a config file is never checked. Wouldn't have to be a big test config, just at least two options to make sure that a newline gets placed between the options. Maybe a test could also verify the indentation of comments on nested sections and options inside those sections.

File: commandapi-platforms/commandapi-velocity/commandapi-velocity-plugin/src/main/java/dev/jorel/commandapi/config/VelocityConfigurationAdapter.java [hunk 4]
```diff
@@ -79,58 +78,28 @@ public CommentedConfigurationNode config() {
...
+
+	@Override
+	public ConfigurationAdapter<ConfigurationNode, DefaultVelocityConfig> loadFromFile(File file) throws IOException {
+		return new VelocityConfigurationAdapter(loader, loader.load(), createDefaultConfig());
+	}
+
+	@Override
+	public void saveToFile(File file) throws IOException {
+		loader.save(config);
 	}
```
[the remark this answers is not in the document]
Comment (DerEchtePilz) [reply commandapi-platforms/commandapi-velocity/commandapi-velocity-plugin/src/main/java/dev/jorel/commandapi/config/VelocityConfigurationAdapter.java:103]: Yeah, ideally I wouldn't have to add the file as a parameter to those methods, however Bukkit works that way so the file parameter is only there for Bukkit since Velocity can just use the loader's file. Really wanted to to avoid to create a loader twice which was criticism of yours before.

**Comment by DerEchtePilz** [comment]
> Still some parts of the code untested: the branches in `BukkitConfigurationAdapter#tryCreateSection`, the `saveDefaultConfig` method, and the rest is stuff `saveDefaultConfig` calls. Those methods look good but the edge cases are important for the functionality, so I think it would be nice to get them covered in tests.

I generally agree, but the `tryCreateSection` method is not too important now since we have one section and I honestly don't see that changing in the near future plus the features that are supposed to work, do work which is verified in manual tests.
`saveDefaultConfig` in my opinion also doesn't need tests since that is functionality that is easily verified in manual tests which did happen. Same with the newlines between options, easily verifiable in manual tests.

Before merging, I am also definitely going to do those manual tests a last time so that shouldn't be problem.

File: commandapi-platforms/commandapi-velocity/commandapi-velocity-plugin/src/main/java/dev/jorel/commandapi/config/VelocityConfigurationAdapter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (willkroboth) [reply commandapi-platforms/commandapi-velocity/commandapi-velocity-plugin/src/main/java/dev/jorel/commandapi/config/VelocityConfigurationAdapter.java:103]: > Another alternative: instead of passing the `File directory` and `File configFile` to `saveDefaultConfig`, those could perhaps be set when constructing `BukkitConfigurationAdapter` and `VelocityConfigurationAdapter`. So neither `loadFromFile` nor `saveToFile` would need a `File` parameter. `VelocityConfigurationAdapter` is already doing this, so this would be the option of updating `BukkitConfigurationAdapter` to match the implementations here.

So maybe this option then? Like currently when `VelocityConfigurationAdapter` is constructed, it is given a `YamlConfigurationLoader` which represents the `File` being edited. On Bukkit it could do a similar thing where I think it would just store the `File` directly as a field and use that in `loadFromFile` and `saveToFile`.


> Really wanted to to avoid to create a loader twice which was criticism of yours before.

Yea, I guess I in this comment https://github.com/JorelAli/CommandAPI/pull/596#discussion_r1791795600 I was more so looking at how `VelocityConfigurationAdapter` and `CommandAPIMain` were using identical code to build their `YamlConfigurationLoader`. If we updated the settings in one on those places, we'd have to remember to update it in the other place. I wouldn't expect creating a `YamlConfigurationLoader` is an especially expensive task that should be avoided, as long as we are sure they are being created the same way each time, which is what I was thinking with the static `createLoader` method.

But yeah, I think there are many possible ways to make the method behavior here consistent between the platforms.

**Review by willkroboth — APPROVED** [review]
> doesn't need tests since that is functionality that is easily verified in manual tests

:shrug: Manual tests are fine, but it sounds tedious to me to redo them each time the relevant code changes. If someone else were to change this code, they might not know that the manual tests need to happen, or they might not handle all cases. That seems like the point of automated tests to me.

Granted, a lot of the code in the CommandAPI is uncovered since it was written before there was a testing framework. But, I like it when new changes are covered as to not make extra work to go back to later.
